### ISSUE TITLE
In a Directory.Build.props file, determine the Sdk value

### ISSUE THREAD

**Issue by Pilchie** [dotnet/msbuild#5255]
_From @kieronlanning on Tuesday, April 7, 2020 1:35:27 PM_

In a `Directory.Build.props` file how can I determine if the project has the `Microsoft.NET.Sdk.Web` or the `Microsoft.NET.Sdk` referenced?

Aside from opening the entire file and using some Regex to pull the `Sdk` attribute value out from either the Project element, or an Import element and setting a property, I'm not sure of the best direction to take...

Currently I'm doing this, which isn't an ideal solution:

```
<PropertyGroup>
 <SdkProjectName>$([System.Text.RegularExpressions.Regex]::Match(`$([System.IO.File]::ReadAllText(`$(MSBuildProjectFullPath)`))`, `(?s-i)(?:^|\s|>)(?s-i)(?:^|\s|&gt;)&lt;\s*(?:Project|Import)\s(?:[^&gt;]*?)\s?Sdk\s*=&quot;(?&lt;sdkproj&gt;.*?)&quot;`).Groups['sdkproj'].Value)</SdkProjectName>
</PropertyGroup>    
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 275b32fd-9419-e174-6fdf-75a0e8ab3a71
* Version Independent ID: f210fd34-23d8-b573-ce6f-9df02e49d80f
* Content: [ASP.NET Core Web SDK](https://docs.microsoft.com/en-us/aspnet/core/razor-pages/web-sdk?toc=%2Fdotnet%2Fcore%2Ftoc.json&bc=%2Fdotnet%2Fbreadcrumb%2Ftoc.json&view=aspnetcore-3.1)
* Content Source: [aspnetcore/razor-pages/web-sdk.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/razor-pages/web-sdk.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-razorpages**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

_Copied from original issue: dotnet/aspnetcore#20649_

**Comment by kieronlanning** [comment]
I've some some interesting properties thanks to the `dotnet msbuild -pp:fullproject.xml {project-file}.csproj`, like `UsingMicrosoftNETSdkWeb` and `UsingMicrosoftNETSdk`.

`UsingMicrosoftNETSdkWeb` looked promising, but it turns it's on true even when a project is using 'Microsoft.NET.Sdk'. Most likely because it references a project that is using `Microsoft.NET.Sdk.Web` (think a test project for example).

**Comment by rainersigwald** [comment]
What do you want to do with this information? There's no easy way to access global project state, intentionally, and this would be something like that. `UsingMicrosoftNETSdkWeb` and friends are indeed the suggested solution.

> `UsingMicrosoftNETSdkWeb` looked promising, but it’s set you true, even when a project is using `Microsoft.NET.Sdk`. Most likely because it references a project that is using `Microsoft.NET.Sdk.Web` (think a test project for example).

Can you give an example? I tried with a trivial `dotnet new xunit` pointing to a `dotnet new web` and it didn't have `UsingMicrosoftNETSdkWeb` in the test project.

**Comment by kieronlanning** [comment]
@rainersigwald We have some quite involved `Directory.Build.props` and `Directory.Build.targets` that automate a lot of the cruft about creating new microservices projects.

Doing things like: 

```
  <!-- Add some web project package references -->
  <ItemGroup Condition="'$(IsWebSdkProject)' == 'true'">
    <PackageReference Include="Microsoft.VisualStudio.Azure.Kubernetes.Tools.Targets" Version="$(MSK8sToolsTarget_Version)" />
    <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="$(MSContainerToolsTarget_Version)" />
  </ItemGroup>
```

Making sure `appsetting.*.json` files are correctly copied to the output directory (including fixing a pet-peeve, nesting those `appsetting.*.json` files in non-SDK projects in VS 2019!):

```

  <!-- Try and nest the appsettings when it's not a web sdk project. -->
  <ItemGroup Condition="'$(IsSdkProject)' == 'true' AND '$(IsWebSdkProject)' == 'false'">
    <Content Include=".\appsettings.*.json">
      <DependentUpon>appsettings.json</DependentUpon>
    </Content>
    <Content Include="appsettings*.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
    </Content>
  </ItemGroup>
```

**Comment by idigra** [comment]
@rainersigwald This information is valuable for MSBuild projects that are built in various environments, especially in precense of breaking changes in the platform. In addition to the example provided by @kieronlanning , here are two examples that I'm blocked on:

1. When moving to .NET 8 in one environment of such project, I encountered the breaking change of [deprecation of ```win10-x64``` runtime identifier](https://learn.microsoft.com/en-gb/dotnet/core/compatibility/sdk/8.0/rid-graph) in favor of ```win-x64```. I need to set the runtime identifier in an early stage of processing (specifically in Directory.Packages.props), and can't just condition by the ```NETCoreSdkVersion``` property which is not available.

2. Analyzers like Microsoft.CodeAnalysis.NetAnalyzers complains when not using an updated version that matches the executing Visual Studio. We used to condition their version by ```NetCoreSdkVersion```. When moving to Central Package Management which is processed earlier, we can't use this property to condition it.
